src/cloud-providers/table.go

```diff
@@ -12,7 +18,68 @@ type CloudProvider interface {
 
 var providerTable map[string]CloudProvider = make(map[string]CloudProvider)
 
+func getSha256sum(providerPath string) (string, error) {
+	file, err := os.Open(providerPath)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	hash := sha256.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return "", err
+	}
+	sum := hash.Sum(nil)
+	return fmt.Sprintf("%x", sum), nil
+}
+
+// LoadCloudProvider loads cloud provider plugin from the given path
+// "${CLOUD_PROVIDER_PLUGIN_PATH}\${CLOUD_PROVIDER}.so"
+// and call the InitCloud function from the found .so files after the sha256sum is verified
+func LoadCloudProvider(name string) {
+	if os.Getenv("CLOUD_PROVIDER_PLUGIN_ENABLE") != "true" {
+		return
+	}
+	cloudProviderPluginPath := os.Getenv("CLOUD_PROVIDER_PLUGIN_PATH")
+	cloudProviderPluginHash := os.Getenv("CLOUD_PROVIDER_PLUGIN_HASH")
+	if cloudProviderPluginPath == "" {
+		logger.Printf("Env CLOUD_PROVIDER_PLUGIN_PATH is not set")
+		return
+	}
+	if cloudProviderPluginHash == "" {
+		logger.Printf("Env CLOUD_PROVIDER_PLUGIN_HASH is not set")
+		return
+	}
+	fullpath := filepath.Join(cloudProviderPluginPath, name+".so")
+	logger.Printf("Loading cloud provider %s from %s", name, fullpath)
+	realPluginHash, err := getSha256sum(fullpath)
+	if err != nil {
+		logger.Printf("Failed to get sha256sum of the cloud provider plugin file %s", err)
+		return
+	}
+	if cloudProviderPluginHash != realPluginHash {
+		logger.Printf("The sha256sum of the cloud provider plugin: %s doesn't match one from configmap: %s", realPluginHash, cloudProviderPluginHash)
+		return
+	}
+	cloud, err := plugin.Open(fullpath)
+	if err != nil {
+		logger.Printf("Failed to open the cloud provider plugin file %s", err)
+	} else {
+		// Every provider plugin must have "InitCloud" function
+		// The code in the function should looks like: "provider.AddCloudProvider("xxx", &Manager{})"
+		ifunc, err := cloud.Lookup("InitCloud")
```

Comment:
Does the golang packages native function `init` called here when plugin.Open? Can we use it directly?

Reply:
Verified that the native `init` function works well in this case, we can use it directly.
Will update the code.